### PR TITLE
Refactor dataprovider

### PR DESCRIPTION
**Pull request by olegsu**
### Summary of your changes
<details><summary>refactor: inject env data provider (c97dea5)</summary> 

This refactor take cares of the dataprovider package.
The data provider interface was change so `FetchData` will get the
resource type and the resource id as well and the return type is now
concrete type instead of an interface, and the `EnrichEvent` is now part
of the data provider interface.

The interesting part is in
(flavors/posture.go)[https://github.com/elastic/cloudbeat/pull/792/files#diff-fc3709e0663635e41188687fee0307f0194885d7c4af2d1b7cd3ff79833bf6f2R224]
where the data provider is been initiated. The refactor moved all the
client creation and all the required metadata to the root so that the
data provider so not require any non-static dependency
 </details>

### Related Issues
- Related https://github.com/elastic/cloudbeat/issues/759

### Checklist
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added the necessary README/documentation (if appropriate)

### CODE REVIEW

File: dataprovider/providers/aws/aws_data_provider.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+)
+
+const (
+	cloudAccountIdField   = "cloud.account.id"
+	cloudAccountNameField = "cloud.account.name"
+	cloudProviderField    = "cloud.provider"
+	cloudProviderValue    = "aws"
+)
+
+type AWSDataProvider struct {
```
Comment (oren-zohar) [inline dataprovider/providers/aws/aws_data_provider.go:36]: Name starts with the package name

File: dataprovider/providers/aws/aws_data_provider.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+func New(options ...Option) AWSDataProvider {
+	adp := AWSDataProvider{}
+	for _, opt := range options {
+		opt(&adp)
+	}
+	return adp
+}
+
+func (a AWSDataProvider) FetchData(ctx context.Context, resource string, id string) (types.Data, error) {
```
Comment (oren-zohar) [inline dataprovider/providers/aws/aws_data_provider.go:50]: I'd replace it with a [blank identifier](https://go.dev/ref/spec#Blank_identifier) for unused params

```suggestion
func (a AWSDataProvider) FetchData(_ context.Context, _ string, id string) (types.Data, error) {
```

File: dataprovider/providers/aws/aws_data_provider.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	return types.Data{
+		ResourceID: id,
+		VersionInfo: version.CloudbeatVersionInfo{
+			Version: version.CloudbeatVersion(),
+			Policy:  version.PolicyVersion(),
+		},
+	}, nil
+}
+
+func (a AWSDataProvider) EnrichEvent(ctx context.Context, event *beat.Event) error {
```
Comment (oren-zohar) [inline dataprovider/providers/aws/aws_data_provider.go:60]:
```suggestion
func (a AWSDataProvider) EnrichEvent(_ context.Context, event *beat.Event) error {
```

File: dataprovider/providers/aws/aws_data_provider.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
```
Comment (oren-zohar) [inline dataprovider/providers/aws/aws_data_provider.go:1]: not sure the file name should have `aws` prefix, as it's part of `aws` package already

File: dataprovider/providers/aws/aws_data_provider_test.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
```
Comment (oren-zohar) [inline dataprovider/providers/aws/aws_data_provider_test.go:1]: not sure the file name should have `aws` prefix, as it's part of `aws` package already

File: dataprovider/providers/k8s/k8s_data_provider.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+)
+
+const (
+	namespace        = "kube-system"
+	clusterNameField = "orchestrator.cluster.name"
+)
+
+var uuidNamespace = uuid.Must(uuid.FromString("971a1103-6b5d-4b60-ab3d-8a339a58c6c8"))
+
+type K8sDataProvider struct {
```
Comment (oren-zohar) [inline dataprovider/providers/k8s/k8s_data_provider.go:40]: Name starts with the package name

File: dataprovider/providers/k8s/k8s_data_provider.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
```
Comment (oren-zohar) [inline dataprovider/providers/k8s/k8s_data_provider.go:1]: same goes for the `k8s` suffix

File: dataprovider/providers/k8s/k8s_data_provider.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+func New(options ...Option) K8sDataProvider {
+	kdp := K8sDataProvider{}
+	for _, opt := range options {
+		opt(&kdp)
+	}
+	return kdp
+}
+
+func (k K8sDataProvider) FetchData(ctx context.Context, resource string, id string) (types.Data, error) {
```
Comment (oren-zohar) [inline dataprovider/providers/k8s/k8s_data_provider.go:58]:
```suggestion
func (k K8sDataProvider) FetchData(_ context.Context, resource string, id string) (types.Data, error) {
```

File: dataprovider/providers/k8s/k8s_data_provider.go [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+	case fetching.CloudContainerMgmt, fetching.CloudIdentity, fetching.CloudLoadBalancer, fetching.CloudContainerRegistry:
+		id = uuid.NewV5(uuidNamespace, k.clusterID).String()
+	}
+	return types.Data{
+		ResourceID:  id,
+		VersionInfo: k.info,
+	}, nil
+}
+
+func (k K8sDataProvider) EnrichEvent(ctx context.Context, event *beat.Event) error {
```
Comment (oren-zohar) [inline dataprovider/providers/k8s/k8s_data_provider.go:71]:
```suggestion
func (k K8sDataProvider) EnrichEvent(_ context.Context, event *beat.Event) error {
```

File: dataprovider/providers/k8s/option.go [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+	}
+}
+
+func WithConfig(cfg *config.Config) Option {
+	return func(kdp *K8sDataProvider) {
+		kdp.cfg = cfg
+	}
+}
+
+func WithInCluster() Option {
```
Comment (oren-zohar) [inline dataprovider/providers/k8s/option.go:40]: this function is not used, maybe worth adding docs for it and the other options

File: flavors/posture.go [hunk 11]
```diff
@@ -91,15 +99,13 @@ func NewPosture(_ *beat.Beat, cfg *agentconfig.C) (*posture, error) {
 	// namespace will be passed as param from fleet on https://github.com/elastic/security-team/issues/2383 and it's user configurable
 	resultsIndex := config.Datastream("", config.ResultsDatastreamIndexPrefix)
 
-	commonDataProvider := dataprovider.NewCommonDataProvider(log, c)
-	commonData, err := commonDataProvider.FetchCommonData(ctx)
+	cdp, err := GetCommonDataProvider(ctx, log, *c)
```
Comment (oren-zohar) [inline flavors/posture.go:102]: opinion: I know it's a common practice in go, but I think too many acronyms can hurt the code readability

**Review by oren-zohar — CHANGES_REQUESTED** [review]
Looks really good! small comments

File: dataprovider/providers/k8s/option.go (hunk 10 above)
Comment (olegsu) [reply dataprovider/providers/k8s/option.go:40]: Right, forgot the remove

File: flavors/posture.go (hunk 11 above)
Comment (olegsu) [reply flavors/posture.go:102]: I agree in general, here the use of the variable is very close to the declaration
https://dave.cheney.net/practical-go/presentations/gophercon-singapore-2019.html#_identifier_length

**Review by oren-zohar — APPROVED** [review]

**Pull request merged**